# Battle of the Neighborhoods - Toronto

In [76]:
import pandas as pd
import numpy as np

## Part 1

### Importing Data

In [10]:
import requests

In [11]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 indicates that the connection to the page was successful.

In [12]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [13]:
type(wiki_data), len(wiki_data)

(list, 3)

Since we only need the first table, we only keep that one.

In [14]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

First, let's ignore cells with a borough that is "Not assigned".

In [15]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Next, let's combine all rows with the same "Postal Code" entry.

In [16]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

We also need to make sure that if a cell has a borough but a "Not assigned" neighborhood, then the neighborhood will be the same as the borough.

Hence, let's check how many such examples there are.

In [17]:
df.Neighbourhood.str.count("Not assigned").sum()

0

There seem to be no such examples, so let's continue with resetting the index.

In [18]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

Let's drop the "index" column as well.

In [19]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Finally, we use the .shape method to print the number of rows of our dataframe.

In [20]:
df.shape

(103, 3)

## Part 2

### Installing geocoder

In [1]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


First we import geocoder.

In [21]:
import geocoder # import geocoder

Next we try to follow the approach suggested on the Coursera page for the Week 3 assignment.

In [ ]:
# initialize your variable to None
lat_lng_coords = None

# taking postal code M5G as an example
postal_code = 'M5G'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

However, since the above approach did not respond after a long computation time, we try the alternative suggestion of importing the CSV file from the given URL.

In [22]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

Let's check that the two dataframes (one scraped from Wikipedia, the other imported from the CSV file) have the same shapes.

In [24]:
print("Shape of dataframe from Wikipedia: ", df.shape)
print("Shape of dataframe from CSV file: ", data.shape)

Shape of dataframe from Wikipedia:  (103, 3)
Shape of dataframe from CSV file:  (103, 3)


Further, in order to be able to join the two dataframes, we need to check the types of the "Postal Code" columns in both dataframes (because we will join the dataframes based on that column). 

In [25]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [26]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

It looks okay, the "Postal Code" columns have the same types. So let's join the two dataframes.

In [27]:
joint_dataframes = df.join(data.set_index('Postal Code'), on = 'Postal Code', how = 'inner')
joint_dataframes

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [28]:
joint_dataframes.shape

(103, 5)

## Part 3

As done in the previous lab for New York City, we cluster Toronto based on the similarities of its venue categories using k-means clustering, and we visualize it using the Foursquare API.

In [2]:
pip install geopy

Note: you may need to restart the kernel to use updated packages.


In [29]:
from geopy.geocoders import Nominatim

First, we get the coordinates of Toronto.

In [30]:
address = 'Toronto, Ontario'
geolocator = Nominatim(user_agent = "toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print("The coordinates of Toronto are {}, {}.".format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Next, we produce a map of Toronto.

In [6]:
pip install folium

Note: you may need to restart the kernel to use updated packages.


In [31]:
import folium

In [33]:
map_Toronto = folium.Map(location = [latitude, longitude], zoom_start=11)

# add markers to the map
for latitude, longitude, borough, neighbourhood in zip(joint_dataframes['Latitude'], joint_dataframes['Longitude'], joint_dataframes['Borough'], joint_dataframes['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker([latitude, longitude], radius = 5, popup = label, color = 'blue', fill = True).add_to(map_Toronto)

map_Toronto

Now we initialize the Foursquare API credentials.

In [34]:
CLIENT_ID = 'ZEQ5QTC1Y0YYNXSPPCX4NNYCXXL1U54PS5F0VN33EG2WMCHC'
CLIENT_SECRET = 'X2KCHJNEATOV4KZV44AKVPUHXZ1SCCYQ3KQXU25BGXEQHUPF'
VERSION = '20180605'

The next step is to create a function for obtaining the nearby (e.g. within a distance of 500m) venue categories in Toronto.

In [37]:
def get_nearby_venues(names, latitudes, longitudes, radius = 500):
    list_of_venues = []
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        # create the URL for the API request
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        lng,
        radius
        )
        # the get-request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        # return only the relevant information for every nearby venue
        list_of_venues.append([(
        name,
        lat,
        lng,
        v['venue']['name'],
        v['venue']['categories'][0]['name']) for v in results])
    nearby_venues = pd.DataFrame([item for list_of_venues in list_of_venues for item in list_of_venues])
    nearby_venues.columns = ['Neighborhood',
                            'Neighborhood Latitude',
                            'Neighborhood Longitude',
                            'Venue',
                            'Venue Category']
    return(nearby_venues)

Using the function "get_nearby_venues", we collect the venues for each neighborhood in Toronto.

In [38]:
toronto_venues = get_nearby_venues(joint_dataframes['Neighbourhood'], joint_dataframes['Latitude'], joint_dataframes['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Let's check the shape of the resulting data.

In [39]:
toronto_venues.shape

(1328, 5)

Apparently there are 1328 venues, with the 5 columns as specified in the function "get_nearby_venues". Now let us have a look at the first 5 rows.

In [40]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

Now let's have a look at the venues based on the neighborhoods.

In [45]:
toronto_venues.groupby('Neighborhood').head()

Neighborhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   
1316  Mimico NW, The Queensway West, South of Bloor,...   
1317  Mimico NW, The Queensway West, South of Bloor,...   
1318  Mimico NW, The Queensway West, South of Bloor,...   

      Neighborhood Latitude  Neighborhood Longitude                   Venue  \
0                 43.753259              -79.329656         Brookbanks Park   
1                 43.753259              -79.329656           Variety Store   
2                 43.725882              -79.315572  Victoria Village Arena   
3                 43.725882              -79.315572               Portugril   
4                 43.725882              -79.315572             Tim Hortons   
...                     ...                     ...                     ...   
1314              43.628841              -79.520999              Wingporium   
1315              43.628841              -79.520999        South St. Burger   
1316              43.628841              -79.520999               Dollarama   
1317              43.628841              -79.520999          Healthy Planet   
1318              43.628841              -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1314            Wings Joint  
1315           Burger Joint  
1316         Discount Store  
1317        Supplement Shop  
1318                 Bakery  

[415 rows x 5 columns]

We see that there are 415 elements for each neighborhood. What happens if we look at the maximum venue category?

In [49]:
toronto_venues.groupby('Venue Category').max()

Neighborhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Wine Bar                                     Little Portugal, Trinity   
Wine Shop                                Dufferin, Dovercourt Village   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighborhood Latitude  Neighborhood Longitude  \
Venue Category                                                      
Accessories Store               43.718518              -79.464763   
Adult Boutique                  43.665860              -79.383160   
Airport                         43.737473              -79.394420   
Airport Food Court              43.628947              -79.394420   
Airport Gate                    43.628947              -79.394420   
...                                   ...                     ...   
Wine Bar                        43.653206              -79.400049   
Wine Shop                       43.669005              -79.442259   
Wings Joint                     43.628841              -79.520999   
Women's Store                   43.733283              -79.419750   
Yoga Studio                     43.715383              -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Wine Bar                            Paris Paris Bar  
Wine Shop                                    Macedo  
Wings Joint                              Wingporium  
Women's Store                         Want Boutique  
Yoga Studio                               YogaSpace  

[235 rows x 4 columns]

We see that there are 235 different venue categories.

### One-hot encoding venue categories

We now want to one-hot encode the venue categories. We do this via the pandas get_dummies function.

In [73]:
toronto_venues_cat = pd.get_dummies(toronto_venues[['Venue Category']], prefix = "", prefix_sep = "")
toronto_venues_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1323                  0               0        0                   0   
1324                  0               0        0                   0   
1325                  0               0        0                   0   
1326                  0               0        0                   0   
1327                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1323             0               0                0                 0   
1324             0               0                0                 0   
1325             0               0                0                 0   
1326             0               0                0                 0   
1327             0               0                0                 0   

      American Restaurant  Antique Shop  ...  Turkish Restaurant  \
0                       0             0  ...                   0   
1                       0             0  ...                   0   
2                       0             0  ...                   0   
3                       0             0  ...                   0   
4                       0             0  ...                   0   
...                   ...           ...  ...                 ...   
1323                    0             0  ...                   0   
1324                    0             0  ...                   0   
1325                    0             0  ...                   0   
1326                    0             0  ...                   0   
1327                    0             0  ...                   0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1323                              0                 0                      0   
1324                              0                 0                      0   
1325                              0                 0                      0   
1326                              0                 0                      0   
1327                              0                 0                      0   

      Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                   0         0          0            0              0   
1                   0         0          0            0              0   
2                   0         0          0            0              0   
3                   0         0          0            0              0   
4                   0         0          0            0              0   
...               ...       ...        ...

Now we add the neighborhood to this one-hot encoded dataframe, and move it to the first column.

In [74]:
toronto_venues_cat['Neighborhood'] = toronto_venues['Neighborhood']

# move the 'Neighborhood' column to the first column
first_col = toronto_venues_cat.pop('Neighborhood')
toronto_venues_cat.insert(0, 'Neighborhood', first_col)

toronto_venues_cat.head()

Neighborhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2  Victoria Village                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Turkish Restaurant  \
0                 0                    0  ...                   0   
1                 0                    0  ...                   0   
2                 0                    0  ...                   0   
3                 0                    0  ...                   0   
4                 0                    0  ...                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                0         0          0            0              0   
1                0         0          0            0              0   
2                0         0          0            0              0   
3                0         0          0            0              0   
4                0         0          0            0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 235 columns]

Next, we group by neighborhoods, and calculate the mean venue categories in each neighborhood.

In [75]:
toronto_grouped = toronto_venues_cat.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  \
0              0.0               0.0                 0.00  ...   
1              0.0               0.0                 0.00  ...   
2              0.0               0.0                 0.00  ...   
3              0.0               0.0                 0.00  ...   
4              0.0               0.0                 0.04  ...   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                    0.0              0.0       0.0        0.0          0.0   
1                    0.0              0.0       0.0        0.0          0.0   
2                    0.0              0.0       0.0        0.0          0.0   
3                    0.0              0.0       0.0        0.0          0.0   
4                    0.0              0.0       0.0        0.0          0.0   

   Women's Store  Yoga Studio  
0           0.00          0.0  
1           0.00          0.0  
2           0.00          0.0  
3           0.00          0.0  
4           0.04          0.0  

[5 rows x 235 columns]

Now we define a function for getting the most common venue category.

In [78]:
def get_most_common_venues(row, num):
    '''''
    row: the row to be considered
    num: the number of venues to be returned
    '''''
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num]

Since there are far too many venue categories, let's start with the top 10 in order to cluster the neighborhoods.

In [79]:
num_of_top_venues = 10
indicators = ['st', 'nd', 'rd']

# create columns based on the chosen number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_of_top_venues):
    try:
        columns.append('{}{} most common venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th most common venue'.format(ind+1))
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = get_most_common_venues(toronto_grouped.iloc[ind, :], num_of_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st most common venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                       Bank   
3                                  Bayview Village                       Bank   
4                Bedford Park, Lawrence Manor East                Coffee Shop   

  2nd most common venue 3rd most common venue 4th most common venue  \
0                Lounge          Skating Rink        Breakfast Spot   
1              Pharmacy        Sandwich Place                   Pub   
2           Coffee Shop              Pharmacy            Restaurant   
3    Chinese Restaurant   Japanese Restaurant                  Café   
4    Italian Restaurant        Sandwich Place      Greek Restaurant   

  5th most common venue        6th most common venue 7th most common venue  \
0           College Gym  Eastern European Restaurant             Drugstore   
1    Athletics & Sports                          Gym           Coffee Shop   
2        Ice Cream Shop                  Supermarket         Deli / Bodega   
3           Yoga Studio             Department Store     Electronics Store   
4            Restaurant            Indian Restaurant             Juice Bar   

         8th most common venue 9th most common venue 10th most common venue  
0                   Donut Shop               Dog Run    Distribution Center  
1                  Curling Ice            Donut Shop                Dog Run  
2    Middle Eastern Restaurant        Sandwich Place            Bridal Shop  
3  Eastern European Restaurant             Drugstore             Donut Shop  
4               Breakfast Spot          Liquor Store                Butcher

Now we create the model for clustering the neighborhoods.

In [80]:
from sklearn.cluster import KMeans

In [104]:
# start with number of clusters = 5
k = 5

toronto_grouped_clustered = toronto_grouped.drop('Neighborhood', 1)

k_means = KMeans(n_clusters = k, random_state = 0).fit(toronto_grouped_clustered)
k_means

KMeans(n_clusters=5, random_state=0)

In [105]:
k_means.labels_[0:]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0,
       0, 0, 2, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 2, 0, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 4,
       0, 2, 0, 0, 0, 0, 2])

Add the clustering label column to the top 10 most common venue categories.

In [ ]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', k_means.labels_)

Now join the dataframes toronto_grouped and joint_dataframes on the 'Neighborhood' column in order to add the latitudes and longitudes for each neighborhood, so as to be able to create a plot.

In [107]:
toronto_merged = joint_dataframes
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on = 'Neighbourhood')
toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st most common venue  \
0  43.753259 -79.329656             2.0                  Park   
1  43.725882 -79.315572             0.0          Intersection   
2  43.654260 -79.360636             0.0           Coffee Shop   
3  43.718518 -79.464763             0.0        Clothing Store   
4  43.662301 -79.389494             0.0           Coffee Shop   

    2nd most common venue 3rd most common venue 4th most common venue  \
0       Food & Drink Shop           Yoga Studio          Dance Studio   
1             Pizza Place           Coffee Shop     French Restaurant   
2                  Bakery                  Park        Breakfast Spot   
3  Furniture / Home Store     Accessories Store           Coffee Shop   
4        Sushi Restaurant    College Auditorium                   Bar   

         5th most common venue  6th most common venue 7th most common venue  \
0  Eastern European Restaurant              Drugstore            Donut Shop   
1                 Hockey Arena  Portuguese Restaurant           Yoga Studio   
2                   Restaurant                   Café                   Pub   
3        Vietnamese Restaurant               Boutique           Event Space   
4                     Beer Bar          Smoothie Shop        Sandwich Place   

  8th most common venue  9th most common venue 10th most common venue  
0               Dog Run    Distribution Center         Discount Store  
1    Dim Sum Restaurant       Department Store           Dessert Shop  
2        Chocolate Shop  Performing Arts Venue     Mexican Restaurant  
3             Gift Shop     Miscellaneous Shop                Dog Run  
4         Burrito Place                   Café  Portuguese Restaurant

Next, drop all NaN values from the dataset.

In [108]:
toronto_final = toronto_merged.dropna(subset=['Cluster Labels'])

Now we can plot the clusters on the map.

In [109]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [110]:
map_clusters = folium.Map(location = [latitude, longitude], zoom_start=11)

# set the color scheme for the clusters
x = np.arange(k) # k = number of clusters, set as 5 above
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_final['Latitude'], toronto_final['Longitude'], toronto_final['Neighbourhood'], toronto_final['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster)+1) + '\n' + str(poi), parse_html=True)
    folium.CircleMarker([lat,lon], radius=5, popup=label, color=rainbow[int(cluster-1)], fill=True, fill_color=rainbow[int(cluster-1)]).add_to(map_clusters)

map_clusters

We can check each of our clusters. 

Cluster 1:

In [111]:
toronto_final.loc[toronto_final['Cluster Labels'] == 0, toronto_final.columns[[1] + list(range(5,toronto_final.shape[1]))]]

Borough  Cluster Labels 1st most common venue  \
1          North York             0.0          Intersection   
2    Downtown Toronto             0.0           Coffee Shop   
3          North York             0.0        Clothing Store   
4    Downtown Toronto             0.0           Coffee Shop   
7          North York             0.0                   Gym   
..                ...             ...                   ...   
96   Downtown Toronto             0.0           Coffee Shop   
97   Downtown Toronto             0.0                  Café   
99   Downtown Toronto             0.0   Japanese Restaurant   
100      East Toronto             0.0    Light Rail Station   
102         Etobicoke             0.0        Discount Store   

      2nd most common venue 3rd most common venue 4th most common venue  \
1               Pizza Place           Coffee Shop     French Restaurant   
2                    Bakery                  Park        Breakfast Spot   
3    Furniture / Home Store     Accessories Store           Coffee Shop   
4          Sushi Restaurant    College Auditorium                   Bar   
7                Restaurant           Coffee Shop            Beer Store   
..                      ...                   ...                   ...   
96               Restaurant    Italian Restaurant                  Café   
97               Restaurant           Coffee Shop    Seafood Restaurant   
99              Coffee Shop    Italian Restaurant      Ramen Restaurant   
100              Smoke Shop           Pizza Place            Restaurant   
102       Convenience Store         Burrito Place                Bakery   

     5th most common venue  6th most common venue 7th most common venue  \
1             Hockey Arena  Portuguese Restaurant           Yoga Studio   
2               Restaurant                   Café                   Pub   
3    Vietnamese Restaurant               Boutique           Event Space   
4                 Beer Bar          Smoothie Shop        Sandwich Place   
7      Japanese Restaurant            Art Gallery      Sushi Restaurant   
..                     ...                    ...                   ...   
96                  Bakery                   Park            Beer Store   
97             Art Gallery              Speakeasy            Steakhouse   
99        Sushi Restaurant   Ethiopian Restaurant           Escape Room   
100          Auto Workshop          Burrito Place  Fast Food Restaurant   
102        Supplement Shop           Burger Joint        Sandwich Place   

    8th most common venue  9th most common venue 10th most common venue  
1      Dim Sum Restaurant       Department Store           Dessert Shop  
2          Chocolate Shop  Performing Arts Venue     Mexican Restaurant  
3               Gift Shop     Miscellaneous Shop                Dog Run  
4           Burrito Place                   Café  Portuguese Restaurant  
7             Supermarket                   Café     Italian Restaurant  
..                    ...                    ...                    ...  
96                Butcher                    Pub   Caribbean Restaurant  
97                  Hotel                    Pub                 Bakery  
99      Indian Restaurant             Restaurant               Beer Bar  
100        Farmers Market                Brewery             Comic Shop  
102  Fast Food Restaurant         Hardware Store          Tanning Salon  

[86 rows x 12 columns]

Cluster 2:

In [112]:
toronto_final.loc[toronto_final['Cluster Labels'] == 1, toronto_final.columns[[1] + list(range(5,toronto_final.shape[1]))]]

Borough  Cluster Labels 1st most common venue 2nd most common venue  \
6  Scarborough             1.0  Fast Food Restaurant           Yoga Studio   

  3rd most common venue 4th most common venue        5th most common venue  \
6         Deli / Bodega     Electronics Store  Eastern European Restaurant   

  6th most common venue 7th most common venue 8th most common venue  \
6             Drugstore            Donut Shop               Dog Run   

  9th most common venue 10th most common venue  
6   Distribution Center         Discount Store

Cluster 3:

In [113]:
toronto_final.loc[toronto_final['Cluster Labels'] == 2, toronto_final.columns[[1] + list(range(5,toronto_final.shape[1]))]]

Borough  Cluster Labels 1st most common venue  \
0         North York             2.0                  Park   
21              York             2.0                  Park   
35         East York             2.0          Intersection   
52        North York             2.0                  Park   
61   Central Toronto             2.0              Bus Line   
66        North York             2.0                  Park   
85       Scarborough             2.0            Playground   
91  Downtown Toronto             2.0                  Park   
98         Etobicoke             2.0                  Park   

         2nd most common venue 3rd most common venue 4th most common venue  \
0            Food & Drink Shop           Yoga Studio          Dance Studio   
21               Women's Store                  Pool          Dance Studio   
35                        Park     Convenience Store           Yoga Studio   
52                 Yoga Studio          Dance Studio     Electronics Store   
61                        Park           Swim School           Yoga Studio   
66  Construction & Landscaping     Convenience Store           Yoga Studio   
85                        Park          Intersection           Yoga Studio   
91                  Playground                 Trail           Yoga Studio   
98                        Pool                 River           Yoga Studio   

          5th most common venue        6th most common venue  \
0   Eastern European Restaurant                    Drugstore   
21  Eastern European Restaurant                    Drugstore   
35                Deli / Bodega            Electronics Store   
52  Eastern European Restaurant                    Drugstore   
61  Eastern European Restaurant                    Drugstore   
66                Deli / Bodega            Electronics Store   
85                 Dance Studio  Eastern European Restaurant   
91                  Curling Ice                    Drugstore   
98                  Curling Ice                    Drugstore   

          7th most common venue 8th most common venue 9th most common venue  \
0                    Donut Shop               Dog Run   Distribution Center   
21                   Donut Shop               Dog Run   Distribution Center   
35  Eastern European Restaurant             Drugstore            Donut Shop   
52                   Donut Shop               Dog Run   Distribution Center   
61                   Donut Shop               Dog Run   Distribution Center   
66  Eastern European Restaurant             Drugstore            Donut Shop   
85                    Drugstore            Donut Shop               Dog Run   
91                   Donut Shop               Dog Run   Distribution Center   
98                   Donut Shop               Dog Run   Distribution Center   

   10th most common venue  
0          Discount Store  
21         Discount Store  
35                Dog Run  
52         Discount Store  
61         Discount Store  
66                Dog Run  
85    Distribution Center  
91         Discount Store  
98         Discount Store

Cluster 4:

In [114]:
toronto_final.loc[toronto_final['Cluster Labels'] == 3, toronto_final.columns[[1] + list(range(5,toronto_final.shape[1]))]]

Borough  Cluster Labels 1st most common venue 2nd most common venue  \
57   North York             3.0        Baseball Field           Yoga Studio   
101   Etobicoke             3.0        Baseball Field          Home Service   

    3rd most common venue 4th most common venue 5th most common venue  \
57            Event Space           Escape Room     Electronics Store   
101           Yoga Studio         Deli / Bodega     Electronics Store   

           6th most common venue 7th most common venue 8th most common venue  \
57   Eastern European Restaurant             Drugstore            Donut Shop   
101  Eastern European Restaurant             Drugstore            Donut Shop   

    9th most common venue 10th most common venue  
57                Dog Run    Distribution Center  
101               Dog Run    Distribution Center

Cluster 5:

In [115]:
toronto_final.loc[toronto_final['Cluster Labels'] == 4, toronto_final.columns[[1] + list(range(5,toronto_final.shape[1]))]]

Borough  Cluster Labels 1st most common venue 2nd most common venue  \
64    York             4.0         Jewelry Store     Convenience Store   

   3rd most common venue 4th most common venue 5th most common venue  \
64           Yoga Studio         Deli / Bodega     Electronics Store   

          6th most common venue 7th most common venue 8th most common venue  \
64  Eastern European Restaurant             Drugstore            Donut Shop   

   9th most common venue 10th most common venue  
64               Dog Run    Distribution Center

Thus, we have clustered Toronto's neighborhoods based on venue categories.